// @dix-vision/governance-core/src/index.ts
/**
 * Governance Core System
 * 
 * Implements the immutable governance axioms and constraint system for DIX VISION.
 * This is the authority layer that defines rules, not executes actions.
 * 
 * CRITICAL: Governance defines rules BEFORE execution, never during.
 */

import type {
  GovernanceConstraints,
  OrderIntent,
  RiskLevel,
  SystemHazardEvent,
  HazardType
} from '@dix-vision/shared-types';
import {
  GovernanceConstraintsSchema,
  OrderIntentSchema,
  SystemHazardEventSchema
} from '@dix-vision/shared-types';

// ============================================================================
// IMMUTABLE GOVERNANCE AXIOMS
// ============================================================================

export const GOVERNANCE_AXIOMS = {
  MAX_DRAWDOWN_DEFAULT: 4, // percentage - hard stop
  MAX_DRAWDOWN_OPERATOR_OVERRIDE: 15, // percentage - operator can increase to this
  MAX_LOSS_PER_TRADE: 5, // percentage
  FAIL_CLOSED: true, // no data → no execution
  CREDENTIALS_NEVER_LEAVE_MACHINE: true,
  ALL_EVENTS_DETERMINISTIC: true,
  ALL_EVENTS_REPLAYABLE: true,
} as const;

export const FORBIDDEN_BEHAVIORS = [
  'martingale_strategies',
  'unbounded_leverage',
  'runtime_core_mutation',
  'unauthorized_network_access',
  'silent_state_mutation',
] as const;

export const REQUIRED_BEHAVIORS = [
  'full_event_traceability',
  'deterministic_replay',
  'governance_defined_constraints',
  'pre_approved_emergency_handling',
] as const;

// ============================================================================
// GOVERNANCE CONSTRAINT SYSTEM
// ============================================================================

export class GovernanceEngine {
  private constraints: GovernanceConstraints;
  private emergencyPolicies: Map<HazardType, EmergencyActionPolicy>;

  constructor(constraints?: Partial<GovernanceConstraints>) {
    this.constraints = this.initializeConstraints(constraints);
    this.emergencyPolicies = this.initializeEmergencyPolicies();
  }

  /**
   * Initialize governance constraints with defaults and custom overrides
   */
  private initializeConstraints(custom?: Partial<GovernanceConstraints>): GovernanceConstraints {
    const defaults: GovernanceConstraints = {
      maxDrawdown: GOVERNANCE_AXIOMS.MAX_DRAWDOWN_DEFAULT,
      maxLossPerTrade: GOVERNANCE_AXIOMS.MAX_LOSS_PER_TRADE,
      failClosed: GOVERNANCE_AXIOMS.FAIL_CLOSED,
      forbiddenBehaviors: [...FORBIDDEN_BEHAVIORS],
      requiredBehaviors: [...REQUIRED_BEHAVIORS],
    };

    const merged = { ...defaults, ...custom };
    return GovernanceConstraintsSchema.parse(merged);
  }

  /**
   * Initialize emergency action policies for hazard types
   * These are PRE-COMPUTED and executed without runtime deliberation
   */
  private initializeEmergencyPolicies(): Map<HazardType, EmergencyActionPolicy> {
    const policies = new Map<HazardType, EmergencyActionPolicy>();

    policies.set('websocket_timeout', {
      hazardType: 'websocket_timeout',
      severity: 'high',
      actions: ['CANCEL_ALL_ORDERS', 'HALT_TRADING'],
      autoExecute: true,
      requiresOperatorApproval: false,
    });

    policies.set('exchange_unreachable', {
      hazardType: 'exchange_unreachable',
      severity: 'critical',
      actions: ['CANCEL_ALL_ORDERS', 'CLOSE_POSITIONS', 'HALT_TRADING'],
      autoExecute: true,
      requiresOperatorApproval: false,
    });

    policies.set('stale_market_data', {
      hazardType: 'stale_market_data',
      severity: 'medium',
      actions: ['HALT_NEW_ORDERS'],
      autoExecute: true,
      requiresOperatorApproval: false,
    });

    policies.set('process_crash', {
      hazardType: 'process_crash',
      severity: 'critical',
      actions: ['CANCEL_ALL_ORDERS', 'CLOSE_POSITIONS', 'HALT_TRADING'],
      autoExecute: true,
      requiresOperatorApproval: false,
    });

    policies.set('data_integrity_failure', {
      hazardType: 'data_integrity_failure',
      severity: 'critical',
      actions: ['CANCEL_ALL_ORDERS', 'HALT_TRADING'],
      autoExecute: true,
      requiresOperatorApproval: false,
    });

    policies.set('clock_drift_detected', {
      hazardType: 'clock_drift_detected',
      severity: 'high',
      actions: ['HALT_TRADING'],
      autoExecute: true,
      requiresOperatorApproval: false,
    });

    policies.set('memory_exhaustion', {
      hazardType: 'memory_exhaustion',
      severity: 'high',
      actions: ['HALT_TRADING', 'TRIGGER_GC'],
      autoExecute: true,
      requiresOperatorApproval: false,
    });

    return policies;
  }

  /**
   * Validate an order intent against governance constraints
   * This returns a precomputed constraint set - does NOT block execution
   */
  public validateOrderIntent(intent: OrderIntent): ValidationResult {
    try {
      // Validate schema
      OrderIntentSchema.parse(intent);

      // Check against constraints
      const checks: ConstraintCheck[] = [];

      // Check if behavior is forbidden
      const intentType = intent.metadata?.behavior as string | undefined;
      if (intentType && this.constraints.forbiddenBehaviors.includes(intentType)) {
        checks.push({
          passed: false,
          constraint: 'forbidden_behavior',
          reason: `Behavior '${intentType}' is forbidden by governance policy`,
        });
      }

      // Check fail-closed policy
      if (this.constraints.failClosed) {
        // Additional validation logic would go here
        checks.push({
          passed: true,
          constraint: 'fail_closed',
          reason: 'Fail-closed policy enforced',
        });
      }

      const passed = checks.every(check => check.passed);

      return {
        passed,
        checks,
        constraintSet: this.generateConstraintSet(),
      };
    } catch (error) {
      return {
        passed: false,
        checks: [{
          passed: false,
          constraint: 'schema_validation',
          reason: error instanceof Error ? error.message : 'Unknown error',
        }],
        constraintSet: this.generateConstraintSet(),
      };
    }
  }

  /**
   * Get emergency action policy for a hazard event
   * Returns precomputed policy for instant execution
   */
  public getEmergencyPolicy(hazard: SystemHazardEvent): EmergencyActionPolicy | null {
    const validated = SystemHazardEventSchema.parse(hazard);
    return this.emergencyPolicies.get(validated.type) || null;
  }

  /**
   * Generate constraint set for execution layer
   * This is precomputed and passed to execution - no runtime governance blocking
   */
  private generateConstraintSet(): ConstraintSet {
    return {
      maxDrawdown: this.constraints.maxDrawdown,
      maxLossPerTrade: this.constraints.maxLossPerTrade,
      failClosed: this.constraints.failClosed,
      timestamp: new Date(),
      version: '42.2.0',
    };
  }

  /**
   * Update constraints (operator action only)
   */
  public updateConstraints(updates: Partial<GovernanceConstraints>): void {
    this.constraints = this.initializeConstraints(updates);
  }

  /**
   * Get current constraints (read-only)
   */
  public getConstraints(): Readonly<GovernanceConstraints> {
    return this.constraints;
  }
}

// ============================================================================
// TYPES
// ============================================================================

export interface ValidationResult {
  passed: boolean;
  checks: ConstraintCheck[];
  constraintSet: ConstraintSet;
}

export interface ConstraintCheck {
  passed: boolean;
  constraint: string;
  reason: string;
}

export interface ConstraintSet {
  maxDrawdown: number;
  maxLossPerTrade: number;
  failClosed: boolean;
  timestamp: Date;
  version: string;
}

export interface EmergencyActionPolicy {
  hazardType: HazardType;
  severity: RiskLevel;
  actions: EmergencyAction[];
  autoExecute: boolean;
  requiresOperatorApproval: boolean;
}

export type EmergencyAction =
  | 'CANCEL_ALL_ORDERS'
  | 'CLOSE_POSITIONS'
  | 'HALT_TRADING'
  | 'HALT_NEW_ORDERS'
  | 'TRIGGER_GC'
  | 'RESTART_SERVICE';
