// @dix-vision/dyon/src/index.ts
/**
 * DYON - System Monitoring Agent
 * 
 * DYON operates in SYSTEM REALITY as a SENSOR (not an executor)
 * Responsibilities: system health monitoring, exchange connectivity tracking, anomaly detection
 * 
 * CRITICAL RULE:
 * "Dyon is a SENSOR. Not an executor."
 * 
 * LIMITS (STRICT):
 * - Cannot execute trades
 * - Cannot call trading APIs
 * - Cannot place/cancel orders
 * - Cannot make market decisions
 * 
 * NEW CAPABILITY:
 * Dyon emits SYSTEM_HAZARD_EVENT for emergency response
 */

import type {
  SystemHazardEvent,
  HazardType,
  SystemHealth,
  ComponentHealth,
  SystemStatus,
  SystemEvent
} from '@dix-vision/shared-types';
import { AGENT_CONFIG } from '@dix-vision/shared-config';

// ============================================================================
// DYON AGENT CLASS
// ============================================================================

export class DyonAgent {
  private config = AGENT_CONFIG.dyon;
  private healthCheckInterval: NodeJS.Timeout | null = null;
  private systemHealth: SystemHealth;
  private eventBus: EventBus;

  constructor(eventBus: EventBus) {
    this.eventBus = eventBus;
    this.systemHealth = this.initializeSystemHealth();
    this.initialize();
  }

  private initialize(): void {
    console.log('DYON Agent initialized');
    console.log(`Health check interval: ${this.config.healthCheckIntervalMs}ms`);
    console.log(`Anomaly threshold: ${this.config.anomalyThreshold}`);
  }

  private initializeSystemHealth(): SystemHealth {
    return {
      status: 'operational',
      uptime: 0,
      lastCheck: new Date(),
      components: [],
    };
  }

  // ============================================================================
  // SYSTEM HEALTH MONITORING
  // ============================================================================

  /**
   * Start continuous health monitoring
   */
  public async start(): Promise<void> {
    console.log('DYON Agent starting health monitoring');
    
    this.healthCheckInterval = setInterval(async () => {
      await this.performHealthCheck();
    }, this.config.healthCheckIntervalMs);

    console.log('DYON health monitoring started');
  }

  /**
   * Stop health monitoring
   */
  public async stop(): Promise<void> {
    console.log('DYON Agent stopping health monitoring');
    
    if (this.healthCheckInterval) {
      clearInterval(this.healthCheckInterval);
      this.healthCheckInterval = null;
    }

    console.log('DYON health monitoring stopped');
  }

  /**
   * Perform comprehensive health check
   * This is the core monitoring function of Dyon
   */
  public async performHealthCheck(): Promise<SystemHealth> {
    const startTime = Date.now();
    
    try {
      // Check all system components
      const componentChecks = await Promise.all([
        this.checkProcessHealth(),
        this.checkMemoryHealth(),
        this.checkNetworkHealth(),
        this.checkExchangeConnectivity(),
        this.checkWebSocketHealth(),
        this.checkClockHealth(),
      ]);

      this.systemHealth = {
        status: this.calculateOverallStatus(componentChecks),
        uptime: process.uptime(),
        lastCheck: new Date(),
        components: componentChecks.filter(c => c !== null) as ComponentHealth[],
      };

      // Detect anomalies
      const anomalies = await this.detectAnomalies(this.systemHealth);
      for (const anomaly of anomalies) {
        await this.emitHazardEvent(anomaly);
      }

      console.log(`DYON health check completed in ${Date.now() - startTime}ms`);
      return this.systemHealth;
    } catch (error) {
      console.error('DYON health check failed:', error);
      
      // Emit hazard event for health check failure
      await this.emitHazardEvent({
        type: 'process_crash',
        severity: 'high',
        component: 'dyon',
        timestamp: new Date(),
        details: { error: error instanceof Error ? error.message : 'Unknown error' },
        requiresEmergencyAction: false,
      });

      throw error;
    }
  }

  /**
   * Check process health
   */
  private async checkProcessHealth(): Promise<ComponentHealth | null> {
    try {
      const uptime = process.uptime();
      
      return {
        name: 'process',
        status: uptime > 0 ? 'operational' : 'degraded',
        latency: 0,
        errorRate: 0,
      };
    } catch (error) {
      return {
        name: 'process',
        status: 'degraded',
        errorRate: 1,
        lastError: error instanceof Error ? error.message : 'Unknown error',
      };
    }
  }

  /**
   * Check memory health
   */
  private async checkMemoryHealth(): Promise<ComponentHealth | null> {
    try {
      const memoryUsage = process.memoryUsage();
      const heapUsedMB = memoryUsage.heapUsed / (1024 * 1024);
      const heapTotalMB = memoryUsage.heapTotal / (1024 * 1024);
      const memoryUsagePercent = (heapUsedMB / heapTotalMB) * 100;

      let status: SystemStatus = 'operational';
      if (memoryUsagePercent > 90) {
        status = 'emergency';
      } else if (memoryUsagePercent > 75) {
        status = 'degraded';
      }

      return {
        name: 'memory',
        status,
        errorRate: 0,
      };
    } catch (error) {
      return {
        name: 'memory',
        status: 'degraded',
        errorRate: 1,
        lastError: error instanceof Error ? error.message : 'Unknown error',
      };
    }
  }

  /**
   * Check network health
   */
  private async checkNetworkHealth(): Promise<ComponentHealth | null> {
    try {
      // Implementation would check network connectivity
      return {
        name: 'network',
        status: 'operational',
        latency: 0,
        errorRate: 0,
      };
    } catch (error) {
      return {
        name: 'network',
        status: 'degraded',
        errorRate: 1,
        lastError: error instanceof Error ? error.message : 'Unknown error',
      };
    }
  }

  /**
   * Check exchange connectivity
   */
  private async checkExchangeConnectivity(): Promise<ComponentHealth | null> {
    try {
      // Implementation would ping exchange APIs
      return {
        name: 'exchange_connectivity',
        status: 'operational',
        latency: 0,
        errorRate: 0,
      };
    } catch (error) {
      return {
        name: 'exchange_connectivity',
        status: 'degraded',
        errorRate: 1,
        lastError: error instanceof Error ? error.message : 'Unknown error',
      };
    }
  }

  /**
   * Check WebSocket health
   */
  private async checkWebSocketHealth(): Promise<ComponentHealth | null> {
    try {
      // Implementation would check WebSocket connections
      return {
        name: 'websocket',
        status: 'operational',
        latency: 0,
        errorRate: 0,
      };
    } catch (error) {
      return {
        name: 'websocket',
        status: 'degraded',
        errorRate: 1,
        lastError: error instanceof Error ? error.message : 'Unknown error',
      };
    }
  }

  /**
   * Check clock health (clock drift detection)
   */
  private async checkClockHealth(): Promise<ComponentHealth | null> {
    try {
      // Implementation would check system clock against NTP
      return {
        name: 'clock',
        status: 'operational',
        latency: 0,
        errorRate: 0,
      };
    } catch (error) {
      return {
        name: 'clock',
        status: 'degraded',
        errorRate: 1,
        lastError: error instanceof Error ? error.message : 'Unknown error',
      };
    }
  }

  /**
   * Calculate overall system status from component health
   */
  private calculateOverallStatus(components: (ComponentHealth | null)[]): SystemStatus {
    const validComponents = components.filter(c => c !== null) as ComponentHealth[];
    
    if (validComponents.some(c => c.status === 'emergency')) {
      return 'emergency';
    }
    if (validComponents.some(c => c.status === 'degraded')) {
      return 'degraded';
    }
    if (validComponents.some(c => c.status === 'maintenance')) {
      return 'maintenance';
    }
    
    return 'operational';
  }

  // ============================================================================
  // ANOMALY DETECTION
  // ============================================================================

  /**
   * Detect anomalies in system health
   */
  private async detectAnomalies(health: SystemHealth): Promise<SystemHazardEvent[]> {
    const hazards: SystemHazardEvent[] = [];

    for (const component of health.components) {
      if (component.status === 'emergency' || component.status === 'degraded') {
        const hazard = this.mapComponentToHazard(component);
        if (hazard) {
          hazards.push(hazard);
        }
      }
    }

    return hazards;
  }

  /**
   * Map component health issue to hazard event
   */
  private mapComponentToHazard(component: ComponentHealth): SystemHazardEvent | null {
    const hazardMap: Record<string, HazardType> = {
      'websocket': 'websocket_timeout',
      'exchange_connectivity': 'exchange_unreachable',
      'clock': 'clock_drift_detected',
      'memory': 'memory_exhaustion',
      'process': 'process_crash',
    };

    const hazardType = hazardMap[component.name];
    if (!hazardType) return null;

    const severity = component.status === 'emergency' ? 'critical' : 'high';

    return {
      type: hazardType,
      severity,
      component: component.name,
      timestamp: new Date(),
      details: {
        errorRate: component.errorRate,
        lastError: component.lastError,
      },
      requiresEmergencyAction: severity === 'critical',
    };
  }

  /**
   * Emit hazard event for emergency response
   * This is how Dyon triggers pre-approved emergency policies
   */
  private async emitHazardEvent(hazard: SystemHazardEvent): Promise<void> {
    console.log(`DYON emitting hazard event: ${hazard.type} (${hazard.severity})`);
    
    await this.eventBus.publish('system_hazard', {
      id: `hazard_${Date.now()}`,
      type: 'system_hazard',
      source: 'dyon',
      timestamp: hazard.timestamp,
      data: hazard,
    });
  }

  // ============================================================================
  // SYSTEM STATUS API
  // ============================================================================

  /**
   * Get current system health
   */
  public getSystemHealth(): SystemHealth {
    return this.systemHealth;
  }

  /**
   * Get component health
   */
  public getComponentHealth(componentName: string): ComponentHealth | undefined {
    return this.systemHealth.components.find(c => c.name === componentName);
  }

  /**
   * Check if system is healthy
   */
  public isSystemHealthy(): boolean {
    return this.systemHealth.status === 'operational';
  }

  // ============================================================================
  // EVENT SYSTEM
  // ============================================================================

  /**
   * Subscribe to system events
   */
  public async subscribeToEvents(eventType: string, handler: (event: SystemEvent) => void): Promise<void> {
    await this.eventBus.subscribe(eventType, handler);
  }
}

// ============================================================================
// SUPPORTING TYPES
// ============================================================================

export interface EventBus {
  publish(eventType: string, event: SystemEvent): Promise<void>;
  subscribe(eventType: string, handler: (event: SystemEvent) => void): Promise<void>;
}
